aima-core/src/main/java/aima/core/agent/DTAgent.java

```diff
@@ -0,0 +1,62 @@
+package aima.core.agent;
+
+
+import aima.core.agent.Action;
+import aima.core.agent.Percept;
+import aima.core.agent.impl.AbstractAgent;
+import aima.core.probability.BeliefState;
+
+import java.util.List;
+
+/**
+ * Artificial Intelligence A Modern Approach (3rd Edition): Figure 13.1, page 484.<br>
+ * 
+ * Figure 13.1  A decision-theoretic agent that selects rational actions.
+ * 
+ * @author Ritwik Sharma
+ */
+public abstract class DTAgent extends AbstractAgent {
+	/**
+	 * @param belief_state
+	 *            		 probabilistic beliefs about the current state of the world.
+	 * @param action
+	 *             the agent's action
+	 */
+	BeliefState<Action,Percept> belief_state;
+	Action action;
+	List<Action> action_descriptions;
+
+	public Action perceive(Percept percept) {
+
+		// update belief_state based on action and percept
+		belief_state.update(action,percept);
+		/** 
+		 * calculate outcome probabilities for actions,
+	   	 *		given action descriptions and current belief_state 
+		*/	
+		List<Action> action_probabilities = cal_action_probabilities(action_descriptions,belief_state);
+		/**  

```

Comment:
Where are you gonna store probabilities calculated from this function? You can't store them in the action_probabilities variable as it is a list of actions and hence cannot store probability values.

Reply:
That Came to my mind too. Well, one solution to that is your implementation in AIMA4e using aima.core.util.datastructure.Pair. I am thinking of some other solution. Thanks for pointing out the mistakes. @samagra14

Reply:
I guess that for this to work you will have to introduce an Action object that stores probabilities. Are you thinking of anything else?